example/multistep/keras_train.py

```diff
@@ -0,0 +1,103 @@
+             .config("spark.driver.memory", "2g") \
+             .getOrCreate())
+    ratingsDF = spark.read.parquet(ratings_data)
+
+    (trainingDF, testDF) = ratingsDF.randomSplit([0.8, 0.2], seed=42)
+    trainingDF.cache()
+    testDF.cache()
+
+    mlflow.log_metric("training_nrows", trainingDF.count())
+    mlflow.log_metric("test_nrows", testDF.count())
+
+    print('Training: {0}, test: {1}'.format(trainingDF.count(), testDF.count()))
+
+    userFactors = alsModel.userFactors.selectExpr("id as userId", "features as uFeatures")
+    itemFactors = alsModel.itemFactors.selectExpr("id as movieId", "features as iFeatures")
+    joinedTrainDF = trainingDF.join(itemFactors, on="movieId").join(userFactors, on="userId")
+    joinedTestDF = testDF.join(itemFactors, on="movieId").join(userFactors, on="userId")
+
+    def concat_arrays(*args):
+        return list(chain(*args))
+
+    concat_arrays_udf = udf(concat_arrays, ArrayType(FloatType()))
+
+    concatTrainDF = (joinedTrainDF
+                     .select('userId', 'movieId', concat_arrays_udf(col("iFeatures"), col("uFeatures")).alias("features"),
+                             col('rating').cast("float")))
+    concatTestDF = (joinedTestDF
+                    .select('userId', 'movieId', concat_arrays_udf(col("iFeatures"), col("uFeatures")).alias("features"),
+                            col('rating').cast("float")))
+
+    pandasDF = concatTrainDF.toPandas()
+    pandasTestDF = concatTestDF.toPandas()
+
+    pandasTestDF[['f0','f1','f2','f3','f4','f5','f6','f7','f8','f9','f10','f11','f12','f13','f14','f15',
+              'f16','f17','f18','f19','f20','f21','f22','f23']] = pd.DataFrame(pandasTestDF.features.values.tolist(), index= pandasTestDF.index)
+
+    x_test = pandasTestDF.drop(["features", "userId", "movieId", "rating"], axis=1)
+
+    pandasDF[['f0','f1','f2','f3','f4','f5','f6','f7','f8','f9','f10','f11','f12','f13','f14','f15',
+              'f16','f17','f18','f19','f20','f21','f22','f23']] = pd.DataFrame(pandasDF.features.values.tolist(), index= pandasDF.index)
+
+    x_train = pandasDF.drop(["features", "userId", "movieId", "rating"], axis=1)
+
+    model = Sequential()
+    model.add(Dense(30, input_dim=24, activation='relu'))
+    model.add(Dense(20, activation='relu'))
+    model.add(Dense(1, activation='linear'))
+
+    model.compile(loss="mse", optimizer=keras.optimizers.Adam(lr=.0001))
+
+    filepath = '/tmp/ALS_checkpoint_weights.hdf5'
+    checkpointer = ModelCheckpoint(filepath=filepath, verbose=1, save_best_only=True)
+    earlyStopping = EarlyStopping(monitor='val_loss', min_delta=0.0001, patience=2, mode='auto')
+
+    history = model.fit(x_train, pandasDF["rating"], validation_split=.2, verbose=2, epochs=3, batch_size=128, shuffle=False, callbacks=[earlyStopping, checkpointer])
+
+    trainMse = model.evaluate(x_train, pandasDF["rating"], verbose=2)
+    testMse = model.evaluate(x_test, pandasTestDF["rating"], verbose=2)
```

Comment:
It might be better to use snake_case for temp variable names throughout, but it's not a huge issue. This particular file mixes snake and camel case though so that's a bit weird.

Reply:
Yes, we can blame @brookewenig for that inconsistency... I can recase everything.